### PR TITLE
Issues/#116 Convert network to SHA-3 (part 2)

### PR DESCRIPTION
**Pull request by macterra**
Re-opened #116 to fix an oversight

### CODE REVIEW

File: src/net.h [hunk 1]
```diff
@@ -555,7 +555,11 @@ class CNodeStats
 
 class CNetMessage {
 private:
+#ifdef BUILD_BTC
     mutable CHash256 hasher;
+#else // BUILD_EQB
+    mutable CSHA3Hash256 hasher;
```
Comment (macterra) [inline src/net.h:561]: This conversion was missed in #139

**Comment by macterra** [comment]
@ealefad All the unit tests pass but it seems the functional tests can't even start... https://travis-ci.com/Equibit/equibit-core/jobs/151511710

Feel free to work directly in this branch.

**Comment by ealefad** [comment]
> @ealefad All the unit tests pass but it seems the functional tests can't even start... https://travis-ci.com/Equibit/equibit-core/jobs/151511710
> 
> Feel free to work directly in this branch.

FT framework can't start the tests since the 2 nodes can't be connected to each other properly, when the commands to connect are sent:
```
./equibit-cli -datadir=/tmp/000/node0 -rpcport=18331 -rpcuser=equibit -rpcpassword=equibit addnode '127.0.0.1:17332' onetry

./equibit-cli -datadir=/tmp/000/node1 -rpcport=18332 -rpcuser=equibit -rpcpassword=equibit addnode '127.0.0.1:17331' onetry
```
it looks like the connection is behaving as uni-directional (check `lastsend` & `lastrecv` values):
```
alex@ubuntu:~/Projects/equibit_dev/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node0 -rpcport=18331 -rpcuser=equibit -rpcpassword=equibit getpeerinfo
[
  {
    "id": 0,
    "addr": "127.0.0.1:17332",
    "addrbind": "127.0.0.1:60998",
    "services": "0000000000000000",
    "relaytxes": false,
    "lastsend": 1539629439,
    "lastrecv": 0,
    "bytessent": 137,
    "bytesrecv": 0,
    "conntime": 1539629439,
    "timeoffset": 0,
    "version": 0,
    "subver": "",
    "inbound": false,
    "addnode": true,
    "startingheight": -1,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
      "version": 137
    },
    "bytesrecv_per_msg": {
    }
  },
  {
    "id": 1,
    "addr": "127.0.0.1:47350",
    "addrbind": "127.0.0.1:17331",
    "services": "0000000000000000",
    "relaytxes": false,
    "lastsend": 0,
    "lastrecv": 1539629455,
    "bytessent": 0,
    "bytesrecv": 137,
    "conntime": 1539629455,
    "timeoffset": 0,
    "version": 0,
    "subver": "",
    "inbound": true,
    "addnode": false,
    "startingheight": -1,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
    },
    "bytesrecv_per_msg": {
      "version": 137
    }
  }
]
```
compare "lastsend" & "lastrecv" taken from the `dev` node :
```
alex@ubuntu:~/Projects/equibit_dev/equibit-core/src$ ./equibit-cli -datadir=/tmp/000/node0 -rpcport=18331 -rpcuser=equibit -rpcpassword=equibit getpeerinfo
[
  {
    "id": 0,
    "addr": "127.0.0.1:17332",
    "addrbind": "127.0.0.1:60936",
    "services": "000000000000040d",
    "relaytxes": true,
    "lastsend": 1539629091,
    "lastrecv": 1539629091,
    "bytessent": 592,
    "bytesrecv": 568,
    "conntime": 1539628851,
    "timeoffset": 0,
    "pingtime": 0.000436,
    "minping": 0.000398,
    "version": 70015,
    "subver": "/Satoshi:0.16.2(testnode1)/",
    "inbound": false,
    "addnode": true,
    "startingheight": 0,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
      "feefilter": 32,
      "getaddr": 24,
      "getheaders": 93,
      "ping": 96,
      "pong": 96,
      "sendcmpct": 66,
      "sendheaders": 24,
      "verack": 24,
      "version": 137
    },
    "bytesrecv_per_msg": {
      "feefilter": 32,
      "getheaders": 93,
      "ping": 96,
      "pong": 96,
      "sendcmpct": 66,
      "sendheaders": 24,
      "verack": 24,
      "version": 137
    }
  },
  {
    "id": 1,
    "addr": "127.0.0.1:47292",
    "addrbind": "127.0.0.1:17331",
    "services": "000000000000040d",
    "relaytxes": true,
    "lastsend": 1539629106,
    "lastrecv": 1539629106,
    "bytessent": 347,
    "bytesrecv": 371,
    "conntime": 1539629105,
    "timeoffset": 0,
    "pingtime": 0.000597,
    "minping": 0.000597,
    "version": 70015,
    "subver": "/Satoshi:0.16.2(testnode1)/",
    "inbound": true,
    "addnode": false,
    "startingheight": 0,
    "banscore": 0,
    "synced_headers": -1,
    "synced_blocks": -1,
    "inflight": [
    ],
    "whitelisted": false,
    "bytessent_per_msg": {
      "feefilter": 32,
      "ping": 32,
      "pong": 32,
      "sendcmpct": 66,
      "sendheaders": 24,
      "verack": 24,
      "version": 137
    },
    "bytesrecv_per_msg": {
      "feefilter": 32,
      "getaddr": 24,
      "ping": 32,
      "pong": 32,
      "sendcmpct": 66,
      "sendheaders": 24,
      "verack": 24,
      "version": 137
    }
  }
]
```
It seems that the nodes can establish "signalling" connection but "traffic" connection is silent.
I'll continue further investigation.

**Comment by ealefad** [comment]
problem in the branch `issues/#116-sha3-net-2` appeared since commit `642abdfeeffd473eea1f392cd4c40eafcc3cb9f8`

**Comment by macterra** [comment]
The core is now expecting a SHA-3 checksum in received message headers:
```
    // Checksum
    CDataStream& vRecv = msg.vRecv;
    const uint256& hash = msg.GetMessageHash();
    if (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) != 0)
    {
        LogPrint(BCLog::NET, "%s(%s, %u bytes): CHECKSUM ERROR expected %s was %s\n", __func__,
           SanitizeString(strCommand), nMessageSize,
           HexStr(hash.begin(), hash.begin()+CMessageHeader::CHECKSUM_SIZE),
           HexStr(hdr.pchChecksum, hdr.pchChecksum+CMessageHeader::CHECKSUM_SIZE));
        return fMoreWork;
    }
```

Can you find where this is added in the client? If this is done in a low level library then we should likely  abandon this change.

**Comment by macterra** [comment]
You can also try running equibitd with `-debug=net` to get more info in the debug.log. See if you get that CHECKSUM ERROR mentioned in the code above.

**Comment by ealefad** [comment]
Yes, checksum error message on the second node.
Log from the node one after sending RPC request `addnode`:
```
2018-10-16 13:35:50.268677 Received a POST request for / from 127.0.0.1:38340
2018-10-16 13:35:50.268819 ThreadRPCServer method=addnode
2018-10-16 13:35:50.268876 trying connection 127.0.0.1:17332 lastseen=0.0hrs
2018-10-16 13:35:50.269304 Added connection peer=0
2018-10-16 13:35:50.269539 sending version (113 bytes) peer=0
2018-10-16 13:35:50.269946 send version message: version 70015, blocks=0, us=[::]:0, peer=0
2018-10-16 13:36:10.068579 Received a POST request for / from 127.0.0.1:38346
2018-10-16 13:36:10.068676 ThreadRPCServer method=getpeerinfo
2018-10-16 13:36:51.007686 socket closed
2018-10-16 13:36:51.007731 disconnecting peer=0
2018-10-16 13:36:51.007777 socket no message in first 60 seconds, 0 1 from 0
2018-10-16 13:36:51.007797 Cleared nodestate for peer=0
```
and the corresponding log from the second node, it receives the connection request over the Bitcoin protocol from the first node, but checksum verification is failed:
```
alex@ubuntu:/tmp/000/node1/regtest$ tail -f debug.log 
2018-10-16 13:35:50.270047 Added connection peer=0
2018-10-16 13:35:50.270118 connection from 127.0.0.1:36136 accepted
2018-10-16 13:35:50.270187 ProcessMessages(version, 113 bytes): CHECKSUM ERROR expected b72d3ba3 was c87b0184
2018-10-16 13:36:51.007505 socket no message in first 60 seconds, 1 0 from 0
2018-10-16 13:36:51.007545 disconnecting peer=0
2018-10-16 13:36:51.007658 Cleared nodestate for peer=0
```
and a later screenshot with Wireshark:
image

**Comment by ealefad** [comment]
there're still messages with invalid checksum, all 15 tests are failing due to 60 sec timeout while waiting for some type of responses. The debug log from the node:
```
2018-10-24 21:46:24.906650 (mocktime: 2018-08-25 21:46:24)   - Connect postprocess: 0.09ms [0.00s (0.06ms/blk)]
2018-10-24 21:46:24.906660 (mocktime: 2018-08-25 21:46:24) - Connect block: 0.30ms [0.00s (0.17ms/blk)]
2018-10-24 21:46:24.906674 (mocktime: 2018-08-25 21:46:24) Checking mempool with 0 transactions and 0 inputs
2018-10-24 21:46:24.906868 (mocktime: 2018-08-25 21:46:24) AddToWallet 22b331a74e08f3d5c93031313990dca37467f58cbba5a70ebc7479e3bd572647  new
2018-10-24 21:46:24.907541 (mocktime: 2018-08-25 21:46:24) AddToWallet 9ae3bb4af43c7965f2d230fa48ada65510337f8ed8400ebc4edc2a1bb89a6dbb  new
2018-10-24 21:46:25.853839 (mocktime: 2018-08-25 21:46:24) Received a POST request for / from 127.0.0.1:40432
2018-10-24 21:46:25.854014 (mocktime: 2018-08-25 21:46:24) ThreadRPCServer method=getblockheader
2018-10-24 21:46:27.008370 (mocktime: 2018-08-25 21:46:24) ProcessMessages(headers, 406 bytes): CHECKSUM ERROR expected f6563a57 was ee7c59fa
2018-10-24 21:46:28.068510 (mocktime: 2018-08-25 21:46:24) ProcessMessages(block, 144 bytes): CHECKSUM ERROR expected d8ef2fa8 was b7c9ab2f
2018-10-24 21:46:28.069271 (mocktime: 2018-08-25 21:46:24) ProcessMessages(ping, 8 bytes): CHECKSUM ERROR expected 084432f3 was 08533f6b
2018-10-24 21:47:23.040051 (mocktime: 2018-08-25 21:46:24) socket no message in first 60 seconds, 1 0 from 0
2018-10-24 21:47:23.040193 (mocktime: 2018-08-25 21:46:24) disconnecting peer=0
2018-10-24 21:47:23.040275 (mocktime: 2018-08-25 21:46:24) Cleared nodestate for peer=0
```
affected messages are `version`, `verack`, `headers`, `block`, `ping` etc. For instance, the expected reply for `version` and `verack` is something like this (from good dev node):
```
should be 2018-10-24 22:21:41.699820 sending version (113 bytes) peer=0
2018-10-24 22:21:41.699939 send version message: version 70015, blocks=0, us=[::]:0, peer=0
2018-10-24 22:21:41.699955 sending verack (0 bytes) peer=0
2018-10-24 22:21:41.700004 receive version message: /python-mininode-tester:0.0.3/: version 70014, blocks=-1, us=127.0.0.1:13308, peer=0
2018-10-24 22:21:41.700030 added time data, samples 2, offset +0 (+0 minutes)
2018-10-24 22:21:41.701507 received: verack (0 bytes) peer=0
```
while in the same test we now stuck at `version` message (debug from the broken node):
```
alex@ubuntu:/tmp/test9hk2yj60$ tail -f node0/regtest/debug.log 
2018-10-24 22:23:39.498121 Received a POST request for / from 127.0.0.1:38964
2018-10-24 22:23:39.498223 ThreadRPCServer method=waitforblockheight
2018-10-24 22:23:39.498991 Received a POST request for / from 127.0.0.1:38964
2018-10-24 22:23:39.499041 ThreadRPCServer method=getrawmempool
2018-10-24 22:23:39.499727 Received a POST request for / from 127.0.0.1:38964
2018-10-24 22:23:39.499759 ThreadRPCServer method=syncwithvalidationinterfacequeue
2018-10-24 22:23:39.507243 Added connection peer=0
2018-10-24 22:23:39.507301 connection from 127.0.0.1:59142 accepted
2018-10-24 22:23:41.471549 Flushing wallet.dat
2018-10-24 22:23:41.472493 Flushed wallet.dat 1ms
2018-10-24 22:24:05.418052 ProcessMessages(version, 116 bytes): CHECKSUM ERROR expected 70f2f6db was 23510bc9
2018-10-24 22:24:40.005290 socket no message in first 60 seconds, 1 0 from 0
2018-10-24 22:24:40.005347 disconnecting peer=0
2018-10-24 22:24:40.005425 Cleared nodestate for peer=0
```

**Review by ealefad — APPROVED** [review]

**Review by r-soltani — APPROVED** [review]

**Pull request merged**